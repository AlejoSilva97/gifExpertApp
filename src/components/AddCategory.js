import React, { useState } from 'react'
import PropTypes from 'prop-types';

//Al venir props.setCategories desestructuramos y sacamos directamente {setCategories}
export const AddCategory = ({setCategories}) => {

    const [inputValue, setinputValue] = useState('');

    //En e.target.value se guarda el valor actual que tiene el input
    const handleInputChange =(e) =>{
        setinputValue(e.target.value);
    }

    const handleSubmit = (e) => {
        //con este metodo prevenimos que se recargue el navegador al hacer submit
        e.preventDefault();

        if ( inputValue.trim().length > 2 ) {
            //Este metodo setCategories se envia como propiedad desde el elemento padre
            setCategories( cats => [inputValue, ...cats]);
            setinputValue('');
        }
    }

    return (
        <form onSubmit={handleSubmit} >
            <p> {inputValue} </p>
            <input 
                type="text"
                value={inputValue}
                onChange={handleInputChange}
            />
        </form>
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
