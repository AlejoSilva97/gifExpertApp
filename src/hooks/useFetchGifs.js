import { useEffect, useState } from "react"
import {getGifs} from '../helpers/getGifs'
 
export const useFetchGifs = (category) => {

    const [state, setstate] = useState({
        data: [],
        loading: true
    });

    //Cuando el segundo parametro es un arreglo vacio el useEffect me permite ejecutar lo que este adentro 
    //la primera vez que se renderize el componente y no cada que detecte cambios en un estado y, 
    //si el arreglo tiene algo, ejecutara lo que este adentro cada que lo que hay en el arreglo cambie.
    useEffect(() => {
        
        //getGifs nos regresa una promesa con las imagenes
        getGifs(category)
            .then((images) => {
                    setstate({
                        data: images,
                        loading: false
                    });
            });

    }, [ category ])

    //Si llamamos la funcion getGifs() en este nivel esta llamara la funcion setImages() 
    //y se volvera a renderizar el componete ya que detecta cambios en un estado, para evitar esto 
    //usamos useEffect() que nos permite llamar la funcion getGifs() solo la primera vez que se renderiza el componente
    //y asi evitamos que se vuelva a ejecutar el llamado a la API
    
    //getGifs();

    return state;
}
