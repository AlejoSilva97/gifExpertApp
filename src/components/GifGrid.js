import React from 'react'
import { useFetchGifs } from '../hooks/useFetchGifs'
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({category}) => {

    //Este es un custom hook creado por nosotros mismos
    const {data:images, loading} = useFetchGifs(category);

    return (
        <>
            {/* Estas clases son animaciones de la libreria animate.css */}
            <h3 className="animate__animated animate__fadeIn">{category}</h3>

            {/* Si loading esta true entonces ponemos un parrafo que indica que esta cargando las imagenes */}
            {loading && <p className="animate__animated animate__flash">Loading...</p>}

            <div className="card-grid">
                {
                    images.map((img)=>(
                        <GifGridItem
                            key={img.id} 
                            {...img}
                        />
                    ))
                }
            </div>
        </>
    )
}
