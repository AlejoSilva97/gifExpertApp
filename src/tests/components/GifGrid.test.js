import React from 'react';
import { shallow } from "enzyme";
import { GifGrid } from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');

describe('Pruebas en el componente <GifGrid />', () => {
    
    const category = 'Naruto';
   
    test('Debe mostrar <GifGrid /> correctamente', () => {

        //Asi simulamos un estado de nuestro useFetchGifs
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });
        
        const wrapper = shallow(<GifGrid category={category}/>);
        expect(wrapper).toMatchSnapshot();

    });
    
    test('Debe de mostrar items cuando se cargan imagenes useFetchGifs', () => {

        const gifs = [{
            id: '123ABC',
            title: 'Naruto Gif',
            url: 'https://cualquiercosa.jpg'
        },
        {
            id: '456DFG',
            title: 'Naruto Gif',
            url: 'https://cualquiercosa.jpg'
        }];

        //Asi simulamos un estado de nuestro useFetchGifs
        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });

        //Un mock se usa para fingir hacer algo
        const wrapper = shallow(<GifGrid category={category}/>);

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('p').exists()).toBe(false);
        expect(wrapper.find('GifGridItem').length).toBe(gifs.length);

    });
    
});
