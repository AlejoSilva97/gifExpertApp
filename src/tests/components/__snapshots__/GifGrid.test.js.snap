// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <GifGrid /> Debe de mostrar items cuando se cargan imagenes useFetchGifs 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    Naruto
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="123ABC"
      key="123ABC"
      title="Naruto Gif"
      url="https://cualquiercosa.jpg"
    />
    <GifGridItem
      id="456DFG"
      key="456DFG"
      title="Naruto Gif"
      url="https://cualquiercosa.jpg"
    />
  </div>
</Fragment>
`;

exports[`Pruebas en el componente <GifGrid /> Debe mostrar <GifGrid /> correctamente 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    Naruto
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
